### PR TITLE
Optionally wait for repo signatures before promoting

### PR DESCRIPTION
**Pull request by puiterwijk**
Signed-off-by: Patrick Uiterwijk <patrick@puiterwijk.org>

### CODE REVIEW

File: bodhi/server/consumers/masher.py
Comment (bowlofeggs) [inline bodhi/server/consumers/masher.py]: Are these ```repo.mash.{complete,failed}``` topics new? I don't see them when git grepping, so I presume they are. It would be good if we add them to these docs as well:

https://fedora-fedmsg.readthedocs.io/en/latest/topics.html#bodhi

File: bodhi/server/consumers/masher.py
Comment (bowlofeggs) [inline bodhi/server/consumers/masher.py]: Optional suggestion: Why not move this if statement to the place where ```wait_for_repo_signature()``` is called? That way, this method always waits if called, which is more like what its name sounds like.

If you like this better I'm not opposed, just a suggestion.

File: bodhi/server/consumers/masher.py
Comment (bowlofeggs) [inline bodhi/server/consumers/masher.py]: This one would be good to add to the fedmsg docs too.

File: bodhi/server/consumers/masher.py
Comment (bowlofeggs) [inline bodhi/server/consumers/masher.py]: Should we have some huge amount of time that we give up and error out after? Like, if it hasn't happened in 10 hours or something arbitrary?

File: bodhi/server/consumers/masher.py
Comment (bowlofeggs) [inline bodhi/server/consumers/masher.py]: This one would be good to document too.

**Review by bowlofeggs — CHANGES_REQUESTED** [review]
Looks pretty good to me, I just have a few suggestions here and there.

The main thing we need is tests. I can help write them sometime soon, but of course feel free to write them too :)

File: bodhi/server/consumers/masher.py
Comment (puiterwijk) [reply bodhi/server/consumers/masher.py]: They are new, yes. And after you agree with the message structure, I would be adding them to fedmsg-meta yes.

File: bodhi/server/consumers/masher.py
Comment (puiterwijk) [reply bodhi/server/consumers/masher.py]: I primarily went with it here because the place where it gets called is really a single series of statements, with just the "if mashing;" conditionals. I'll move it over.

File: bodhi/server/consumers/masher.py
Comment (puiterwijk) [reply bodhi/server/consumers/masher.py]: Well, if we at some point enable this on clients, and we send any repo without signature, that repo will be useless.
So when we turn this on, we should make it required to proceed.

Comment (puiterwijk) [reply bodhi/server/consumers/masher.py]: Also, during the initial time while it's not yet enabled for clients, if for whatever reason the signature doesn't get generated, we can just create the files by hand.

**Comment by puiterwijk** [comment]
Given the massive changes to masher.py in #1900, I'm going to wait for that to get merged, after that I'll write the tests for this.

**Comment by bowlofeggs** [comment]
There are again massive changes in #2052, so you might want to either try to beat it to merge or wait until it merges to pick this back up.

**Comment by bowlofeggs** [comment]
I am going to work on rebasing this. Might not finish it today though.

**Comment by bowlofeggs** [comment]
Alright, I've gotten it rebased, and I also added a new ComposeState for this so we can see if it gets stuck here. I've not yet tested the migration so we might need some adjustments there, and we still need some tests.

I'll be gone for a while, but I hope to return to this when I get back.

File: bodhi/server/consumers/masher.py
Comment (bowlofeggs) [inline bodhi/server/consumers/masher.py]: Are these two messages related to waiting on repo signatures? If not, I'd prefer not to include this change in the patch to keep the patch more targeted. Also, there already is a message sent upon completion and it indicates success/failure, so is this redundant? Lastly, if we do keep this, should we keep using the word "mash" since it's composes now? We can't rename the existing messages without bumping Bodhi's X version, but I've files a [ticket](https://github.com/fedora-infra/bodhi/issues/2151) to do that later.

Comment (puiterwijk) [reply bodhi/server/consumers/masher.py]: So, those two messages can be dropped, but I'll need `self.path` as argument to the `repo.signature.wait` message.

**Comment by bowlofeggs** [comment]
Here's a patch to fix a small issue with the migration and to add the new setting to ```production.ini```:

```
diff --git a/bodhi/server/migrations/versions/c21dd18b161a_add_a_new_signing_repo_composestate_enum.py b/bodhi/server/migrations/versions/c21dd18b161a_add_a_new_signing_repo_composestate_enum.py
index a4901546..32d3eddc 100644                            
--- a/bodhi/server/migrations/versions/c21dd18b161a_add_a_new_signing_repo_composestate_enum.py                       
+++ b/bodhi/server/migrations/versions/c21dd18b161a_add_a_new_signing_repo_composestate_enum.py                       
@@ -23,8 +23,7 @@ Revises: 2616c86d8ac6                    
 Create Date: 2018-01-29 16:18:00.441595                   
 """                         
 from alembic import op      
-import sqlalchemy as sa     
-                            
+from sqlalchemy import exc  
                             
                             
 # revision identifiers, used by Alembic.                  
diff --git a/bodhi/tests/server/consumers/test_masher.py b/bodhi/tests/server/consumers/test_masher.py                
index 80467094..d81ce594 100644                            
--- a/bodhi/tests/server/consumers/test_masher.py          
+++ b/bodhi/tests/server/consumers/test_masher.py          
@@ -1,5 +1,5 @@              
 # -*- coding: utf-8 -*-     
-# Copyright © 2007-2017 Red Hat, Inc.                     
+# Copyright © 2007-2018 Red Hat, Inc.                     
 #                           
 # This file is part of Bodhi.                             
 #                           
diff --git a/production.ini b/production.ini               
index 81dce3c9..cbe60deb 100644                            
--- a/production.ini         
+++ b/production.ini         
@@ -116,6 +116,9 @@ use = egg:bodhi-server                 
 # location of this file.    
 # mash_stage_dir =          
                             
+# Whether to wait for .asc signature files in the repo when composing updates or not.                                
+# wait_for_repo_sig = False 
+                            
 # The following jinja2 template variables are available for use to customize the Pungi configs and                   
 # variants files to the Release and Updates:              
 #
```

File: bodhi/server/models.py
Comment (mscherer) [inline bodhi/server/models.py]: Shouldn't it be "signed" rather than "siged" ?

File: bodhi/server/consumers/masher.py
Comment (mscherer) [inline bodhi/server/consumers/masher.py]: Is it safe to wait basically "forever" and never fail ? 
I could imagine a few race conditions where this would slowly exhaust ressources, so maybe there should be some limit somewhere, unless that's already handled externally.

Comment (bowlofeggs) [reply bodhi/server/consumers/masher.py]: @mscherer Yeah I had a similar thought - we probably want it to fail if it's not signed after some timeout, just so it doesn't sit around forever.

Comment (puiterwijk) [reply bodhi/server/consumers/masher.py]: I would say that if there's something wrong to the point where we can't get a repo signed, releng should just abort the push by hand and disable the option.
The problem is that it might be that there's a backlog for signing up to multiple hours, and then there's some time needed for releng to debug it.
If we at some point decide to turn this on and we start shipping configurations with repo_gpgcheck=1, we just cannot ship unsigned repos anymore.

So, I personally think that we should not have an automated timeout, and any timing out should be done by releng.

**Comment by bowlofeggs** [comment]
On 02/28/2018 08:18 PM, Patrick Uiterwijk wrote:
> If we at some point decide to turn this on and we start shipping
> configurations with repo_gpgcheck=1, we just cannot ship unsigned repos
> anymore.

I think the suggestion is that the compose would fail after the timeout
(i.e., we would not ship unsigned repos to users, we would just mark the
compose as failed automatically.)

**Comment by puiterwijk** [comment]
Right, and I tried to say that even that is not something we'd probably want (or need) I think.
There are not many resources accumulated during the wait, but not having an automated timeout means that we don't inadvertently abort the push just because robosignatory was busy processing a mass rebuild.
If we want the push to fail because of signing issues, releng can do that.

**Comment by puiterwijk** [comment]
Tests have been added.

File: bodhi/server/consumers/masher.py
Comment (bowlofeggs) [inline bodhi/server/consumers/masher.py]: This change is unrelated to the patch at hand, and I also would prefer to keep this line here.

File: bodhi/server/consumers/masher.py
Comment (bowlofeggs) [inline bodhi/server/consumers/masher.py]: Shouldn't we only send this message if we are going to wait (i.e., do this inside the if block below)?

File: bodhi/tests/server/consumers/test_masher.py
Comment (bowlofeggs) [inline bodhi/tests/server/consumers/test_masher.py]: I suggest moving that message into the if block and changing this assertion to:

```
self.assertEqual(publish.call_count, 0)
```

**Review by bowlofeggs — APPROVED** [review]
Only a few small changes - feel free to merge without re-review, and thanks!

**Pull request merged**
